be/dtm0_log.c

```diff
@@ -0,0 +1,298 @@
+
+	M0_ALLOC_PTR(lrec);
+	if (lrec == NULL)
+		return M0_ERR(-ENOMEM);
+
+	m0_dtm0_tx_desc_copy(txd, &lrec->dlr_txd);
+
+	m0_buf_copy(&lrec->dlr_pyld, pyld);
+
+	*rec = lrec;
+	return 0;
+}
+
+static void m0_be_dtm0_log_rec_fini(struct m0_dtm0_log_rec **rec,
+                                    struct m0_be_tx         *tx)
+{
+	struct m0_dtm0_log_rec *lrec = *rec;
+
+	if (lrec->dlr_pyld.b_nob) {
+		m0_buf_free(&lrec->dlr_pyld);
+	}
+
+	m0_dtm0_tx_desc_fini(&lrec->dlr_txd);
+	m0_free(lrec);
+	*rec = NULL;
+}
+
+static int m0_be_dtm0_log_rec__updt_past(enum m0_dtm0_tx_pa_state *dst,
+                                         enum m0_dtm0_tx_pa_state *src)
+{
+	if (*dst < *src) {
+		*dst = *src;
+	}
+
+	return 0;
+}
+
+static int m0_be_dtm0_log__insert(struct m0_be_dtm0_log  *log,
+                                  struct m0_be_tx        *tx,
+                                  struct m0_dtm0_tx_desc *txd,
+                                  struct m0_buf          *pyld)
+{
+	int                     rc;
+	struct m0_dtm0_log_rec *rec = NULL;
+
+	rc = m0_be_dtm0_log_rec_int(&rec, tx, txd, pyld);
+	if (rc)
+		return rc;
+
+	lrec_tlink_init_at_tail(rec, log->dl_tlist);
+	return rc;
+}
+
+static int m0_be_dtm0_log__set(struct m0_be_dtm0_log  *log,
+                               struct m0_be_tx        *tx,
+                               struct m0_dtm0_tx_desc *txd,
+                               struct m0_buf          *pyld,
+                               struct m0_dtm0_log_rec *rec)
+{
+	struct m0_dtm0_tx_desc *ltxd     = &rec->dlr_txd;
+	struct m0_buf          *lpyld    = &rec->dlr_pyld;
+	struct m0_dtm0_tx_pa   *ldtpg_pa = ltxd->dtd_pg.dtpg_pa;
+	struct m0_dtm0_tx_pa   *dtpg_pa  = txd->dtd_pg.dtpg_pa;
+	uint32_t                num_pa   = ltxd->dtd_pg.dtpg_nr;
+
+	M0_PRE(m0_dtm0_log_rec__invariant(rec));
+
+	/* Attach payload to log if it is not attached */
+	if (!lpyld->b_nob && pyld->b_nob) {
+		m0_buf_copy(lpyld, pyld);
+	}
+
+	/* Update participant state if changed */
+	m0_forall(pa_id, num_pa,
+                  !m0_be_dtm0_log_rec__updt_past(&ldtpg_pa[pa_id].pa_state,
+                                                 &dtpg_pa[pa_id].pa_state));
+	return 0;
+}
+
+M0_INTERNAL int m0_be_dtm0_log_update(struct m0_be_dtm0_log  *log,
+                                      struct m0_be_tx        *tx,
+                                      struct m0_dtm0_tx_desc *txd,
+                                      struct m0_buf          *pyld)
+{
+	int                     rc;
+	struct m0_dtm0_log_rec *rec = NULL;
+
+	M0_PRE(pyld);
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tx_desc__invariant(txd));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	rc = m0_be_dtm0_log_find(log, &txd->dtd_id, &rec);
+	if (rc == 0) {
+		rc = m0_be_dtm0_log__set(log, tx, txd, pyld, rec);
+	} else if (rc == -ENOENT) {
+		rc = m0_be_dtm0_log__insert(log, tx, txd, pyld);
+	}
+	return rc;
+}
+
+M0_INTERNAL int m0_be_dtm0_log_prune(struct m0_be_dtm0_log    *log,
```

Comment:
[*logic*] i'd suppose that _prune() is going to delete records from the end to the given one().

Reply:
Done at 65fd240824549185c027ed78a9419cefa0b039ff